### DISCUSSION TITLE
Best practices for authoring middleware

### DISCUSSION THREAD

**Discussion by charkour** [pmndrs/zustand#2305] [Q&A]
As a zustand middleware author, I want to clarify what the best practices are for authoring middleware that modifying `set` and `get`. 

On the main [readme](https://github.com/pmndrs/zustand#middleware), it **only** modifies `set`. But on the TypeScript [readme](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#middleware-that-doesnt-change-the-store-type), it modifies both the `set` and `setState`. I think it is expected that middleware should modify both, but that's not specified. I follow this pattern on [zundo](https://github.com/charkour/zundo/blob/9432887ca8ac14347ca14d5b2b1349fc7cf50076/src/index.ts#L81).

As an aside, I think the example in the TypeScript documentation has a bug because it's using `set` inside the modified `setState` which can cause issues if the user expects the nested middlewares to properly pass modified `set` and `setState` behavior. 

Is it best practice to have middleware modify both `set` and `setState`? Should we specify that on the zustand reamde?

For example, which example below is the best practice?

This one only modifies `set`. 
```ts
const log = (config) => (set, get, api) =>
  config(
    (...args) => {
      console.log('  applying', args)
      set(...args)
      console.log('  new state', get())
    },
    get,
    api,
  )
```

```ts
const log = (config) => (set, get, api) => {
  const loggedSet = (...args) => {
      console.log('  applying', args)
      set(...args)
      console.log('  new state', get())
    }
  api.setState = loggedSet // potential bug if nested middleware modifies setState
  return config(
    loggedSet,
    get,
    api,
  )
}
```

**Comment by charkour** [comment]
My proposed changes: https://github.com/pmndrs/zustand/pull/2306

**Comment by dai-shi** [comment]
I think middleware library should change both.
But, app developer adding custom middleware doesn't have to (because they know they use `setState` or not.)

Having that said, I think we should remove the section about creating custom middleware in readme.md, and just add a link to a guide doc. (because, with TypeScript, it's extremely hard to create a custom middleware.)
